mqtt-simulator/simulator.py

```diff
@@ -11,26 +11,35 @@ def __init__(self, settings_file):
     def load_settings(self, settings_file):
         with open(settings_file) as json_file:
             config = json.load(json_file)
-            self.broker_url = config['BROKER_URL']
-            self.broker_port = config['BROKER_PORT']
+            self.broker_url = config.get('BROKER_URL', 'localhost')
+            self.broker_port = config.get('BROKER_PORT', 1883)
+            broker_time_interval = config.get('TIME_INTERVAL', 10)
+            broker_protocol = config.get('PROTOCOL_VERSION', 4) # mqtt.MQTTv311
+            broker_clean = config.get('CLEAN_SESSION', True)
+            broker_qos = config.get('QOS', 2)
+            broker_retained = config.get('RETAINED', False)
             # read each configured topic
             for topic in config['TOPICS']:
                 topic_data = topic['DATA']
-                topic_time_interval = topic['TIME_INTERVAL']
+                topic_time_interval = topic.get('TIME_INTERVAL', broker_time_interval)
+                topic_protocol = topic.get('PROTOCOL_VERSION', broker_protocol)
+                topic_clean = topic.get('CLEAN_SESSION', broker_clean)
+                topic_qos = topic.get('QOS', broker_qos)
+                topic_retained = topic.get('RETAINED', broker_retained)
```

Comment:
Since we are using these 5 parameters in broker and topic do you think we can create a separate file, like `client_settings.py`, with the dataclass below?

```python
from dataclasses import dataclass

@dataclass
class ClientSettings:
    protocol: int
    qos: int
    clean: bool
    retained: bool
    time_interval: int
```

We can also have a handle / init that receives only the input dictionary and a dictionary of default values for better readability - `topic_client_settings = read_client_settings(topic, broker_client_setting)`.

Reply:
This is a great improvement. The client settings could be defined as a static file and even map to env vars. I would add the ability for the data files to be streamed in through stdin as an option. I am calling this simulator from a golang service and could just feed the `raw_values` json straight into the script.

Reply:
I will address this after this PR.